twingate/internal/provider/resource/resource.go

```diff
@@ -4,444 +4,966 @@ import (
+										listvalidator.SizeAtMost(1),
+									},
+									NestedObject: schema.NestedBlockObject{
+										Attributes: map[string]schema.Attribute{
+											attr.Policy: schema.StringAttribute{
+												Optional: true,
+												Computed: true,
+											},
+											attr.Ports: schema.SetAttribute{
+												Optional:    true,
+												Computed:    true,
+												ElementType: types.StringType,
+											},
+										},
+									},
+								},
+								attr.TCP: schema.ListNestedBlock{
+									Validators: []validator.List{
+										listvalidator.SizeAtMost(1),
+									},
+									NestedObject: schema.NestedBlockObject{
+										Attributes: map[string]schema.Attribute{
+											attr.Policy: schema.StringAttribute{
+												Optional: true,
+												Computed: true,
+											},
+											attr.Ports: schema.SetAttribute{
+												Optional:    true,
+												Computed:    true,
+												ElementType: types.StringType,
+											},
+										},
+									},
+								},
+							},
+						},
+					},
+				},
+			},
+			StateUpgrader: func(ctx context.Context, req resource.UpgradeStateRequest, resp *resource.UpgradeStateResponse) {
+				var priorState resourceModelV0
+
+				resp.Diagnostics.Append(req.State.Get(ctx, &priorState)...)
+
+				if resp.Diagnostics.HasError() {
+					return
+				}
+
+				protocols, err := convertProtocolsV0(priorState.Protocols)
+				if err != nil {
+					resp.Diagnostics.AddError(
+						"failed to convert protocols for prior state version 0",
+						err.Error(),
+					)
+
+					return
+				}
+
+				protocolsState, diags := convertProtocolsToTerraform(protocols, nil)
+				resp.Diagnostics.Append(diags...)
+				if resp.Diagnostics.HasError() {
+					return
+				}
+
+				upgradedState := resourceModel{
+					ID:              priorState.ID,
+					Name:            priorState.Name,
+					Address:         priorState.Address,
+					RemoteNetworkID: priorState.RemoteNetworkID,
+					Protocols:       protocolsState,
+					Access:          priorState.Access,
+				}
+
+				if !priorState.IsAuthoritative.IsNull() {
+					upgradedState.IsAuthoritative = priorState.IsAuthoritative
+				}
+
+				if !priorState.IsVisible.IsNull() {
+					upgradedState.IsVisible = priorState.IsVisible
+				}
+
+				if !priorState.IsBrowserShortcutEnabled.IsNull() {
+					upgradedState.IsBrowserShortcutEnabled = priorState.IsBrowserShortcutEnabled
+				}
+
+				if !priorState.Alias.IsNull() && priorState.Alias.ValueString() != "" {
+					upgradedState.Alias = priorState.Alias
+				}
+
+				if !priorState.SecurityPolicyID.IsNull() && priorState.SecurityPolicyID.ValueString() != "" {
+					upgradedState.SecurityPolicyID = priorState.SecurityPolicyID
+				}
+
+				resp.Diagnostics.Append(resp.State.Set(ctx, upgradedState)...)
+
+				resp.Diagnostics.AddWarning("Please upgrade protocols sections", "Follow this docs to update protocols from blocks to attributes")
```

Comment:
we should add needed info to this warning msg

Reply:
@twingate-blee @alexb-twingate what about this warning message, how should it look like?

Reply:
@alexb-twingate How about 
"Please update the protocols sections format from a block to an object", "There is a protocols object example in the twingate_resource documentation"

Reply:
I like that. I'd change the longer description to "See the v1 to v2 migration guide in the Twingate Terraform Provider documentation" per what we talked about earlier today.

Reply:
updated this message